packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts

```diff
@@ -107,86 +131,78 @@ export class RemoteMessageProcessor {
+				this.batchInProgress === undefined,
 				0x9d3 /* Grouped batch interrupting another batch */,
 			);
-			assert(
-				this.processorBatch.length === 0,
-				0x9d4 /* Processor batch should be empty on grouped batch */,
-			);
+			const batchId = asBatchMetadata(message.metadata)?.batchId;
 			const groupedMessages = this.opGroupingManager.ungroupOp(message).map(unpack);
-			// If the batch is empty, we need to return the sequence number aside
-			if (groupedMessages.length === 0) {
-				assert(
-					message.sequenceNumber !== undefined,
-					"Empty grouped batch has no sequence number",
-				);
-				return {
-					messages: groupedMessages, // empty array
-					batchStartCsn: message.clientSequenceNumber,
-					sequenceNumber: message.sequenceNumber,
-				};
-			}
 			return {
-				messages: groupedMessages,
+				messages: groupedMessages, // Will be [] for an empty batch
 				batchStartCsn: message.clientSequenceNumber,
+				clientId,
+				batchId,
+				// If the batch is empty, we need to return the sequence number aside
+				emptyBatchSequenceNumber:
+					groupedMessages.length === 0 ? message.sequenceNumber : undefined,
 			};
 		}
 
-		const batchStartCsn = this.getAndUpdateBatchStartCsn(message);
-
 		// Do a final unpack of runtime messages in case the message was not grouped, compressed, or chunked
 		unpackRuntimeMessage(message, logLegacyCase);
-		this.processorBatch.push(message as InboundSequencedContainerRuntimeMessage);
 
-		// this.batchStartCsn is undefined only if we have processed all messages in the batch.
-		// If it's still defined, we're still in the middle of a batch, so we return nothing, letting
-		// containerRuntime know that we're waiting for more messages to complete the batch.
-		if (this.batchStartCsn !== undefined) {
+		const batchEnd = this.addMessageToBatch(
+			message as InboundSequencedContainerRuntimeMessage & { clientId: string },
+		);
+
+		if (!batchEnd) {
 			// batch not yet complete
 			return undefined;
 		}
 
-		const messages = [...this.processorBatch];
-		this.processorBatch.length = 0;
-		return {
-			messages,
-			batchStartCsn,
-		};
+		const completedBatch = this.batchInProgress;
+		this.batchInProgress = undefined;
+		return completedBatch;
 	}
 
 	/**
-	 * Based on pre-existing batch tracking info and the current message's batch metadata,
-	 * this will return the starting CSN for this message's batch, and will also update
-	 * the batch tracking info (this.batchStartCsn) based on whether we're still mid-batch.
+	 * Add the given message to the current batch, and indicate whether the batch is now complete.
+	 *
+	 * @returns true if the batch is now complete, false if more messages are expected
 	 */
-	private getAndUpdateBatchStartCsn(message: ISequencedDocumentMessage): number {
+	private addMessageToBatch(
+		message: InboundSequencedContainerRuntimeMessage & { clientId: string },
+	): boolean {
 		const batchMetadataFlag = asBatchMetadata(message.metadata)?.batch;
```

Comment:
I really like how things are more explicit now with batchEnd for example, but this is one which can also improve like that. batchMetadataFlag is not that explicit (metadata.batch = true means end of the batch is worse but probably nothing we can do about it now). I think batchMetadataFlag could be something like "batchEndMetadataFlag" "endOfBatch" and also it could be good to separate addMessageToBatch with determining if batch ended (or call it something superverbose like addMessageToBatchAndIsBatchEnd)

Reply:
Not sure I totally follow.  Is this feedback for this PR, or just in general?  The batchMetadataFlag here indicates batch _start_ if `true` and batch _end_ if `false` and middle of batch or single-message-batch otherwise.  These semantics are baked in several places in the code (written in `BatchManager.popBatch`).

I'll think about the feedback about the return value being surprising relative to the name, good point.

Reply:
I'll give the return type some structure so that boolean is named, like `{ batchEnded: true }`

Reply:
I mean in general, batch start if true and batch end if false is really not that explicit but I know there is little to do about that at this point. I was just pointing out we could improve a bit that being more explicit here which batchEnded helps on that :)

Reply:
Yep agree the existing stuff is kind of cryptic. But it's also the wire format so maybe it's concise for optimizing that. Not sure the history.